src/jdk.crypto.cryptoki/share/classes/sun/security/pkcs11/SunPKCS11.java

```diff
@@ -539,12 +551,58 @@ public long importKey(long hSession, CK_ATTRIBUTE[] attributes) throws PKCS11Exc
             if (attr.type == CKA_KEY_TYPE) {
                 keyType = attr.getLong();
             }
-            if (attr.type == CKA_VALUE) {
-                keyBytes = attr.getByteArray();
+            ckAttrsMap.put(attr.type, attr);
+        }
+
+        if (keyClass == CKO_PRIVATE_KEY) {
+            if (keyType == CKK_RSA) {
+                try {
+                    keyBytes = RSAPrivateCrtKeyImpl.newKey(
+                        KeyType.RSA,
+                        null,
+                        getBigIntegerOrZero(ckAttrsMap, CKA_MODULUS),
+                        getBigIntegerOrZero(ckAttrsMap, CKA_PUBLIC_EXPONENT),
+                        getBigIntegerOrZero(ckAttrsMap, CKA_PRIVATE_EXPONENT),
+                        getBigIntegerOrZero(ckAttrsMap, CKA_PRIME_1),
+                        getBigIntegerOrZero(ckAttrsMap, CKA_PRIME_2),
+                        getBigIntegerOrZero(ckAttrsMap, CKA_EXPONENT_1),
+                        getBigIntegerOrZero(ckAttrsMap, CKA_EXPONENT_2),
+                        getBigIntegerOrZero(ckAttrsMap, CKA_COEFFICIENT)
+                    ).getEncoded();
+                } catch (InvalidKeyException e) {
+                    throw new PKCS11Exception(0x00000005L, null);
+                }
+            } else if (keyType == CKK_EC) {
+                if (!ckAttrsMap.containsKey(CKA_EC_PARAMS)) {
+                    throw new PKCS11Exception(0x00000005L, "CKA_EC_PARAMS attribute is missing.");
+                }
```

Comment:
Suggest this should just `get()` the attribute earlier and use a `null` check.

Reply:
The point of just using `get()` was to avoid two accesses to `ckAttrsMap`; the result of the (first and only) call to `get()` should be captured in a local variable for reuse on line 584.

Reply:
Updated.